internal/webhooks/patch_validation.go

```diff
@@ -0,0 +1,360 @@
+					"jsonPatch path must start with \"/spec/\"",
+				))
+		}
+
+		// Validate that array access is only prepend or append for add and not allowed for replace or remove.
+		allErrs = append(allErrs,
+			validateIndexAccess(jsonPatch, path.Index(i).Child("path"))...,
+		)
+
+		// Validate the value and valueFrom fields for the patch.
+		allErrs = append(allErrs,
+			validateJSONPatchValues(jsonPatch, variableSet, path.Index(i))...,
+		)
+	}
+	return allErrs
+}
+
+func validateJSONPatchValues(jsonPatch clusterv1.JSONPatch, variableSet map[string]*clusterv1.ClusterClassVariable, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	// move to the next variable if the jsonPatch does not have "replace" or "add" op. Additional validation is not needed.
+	if jsonPatch.Op != "add" && jsonPatch.Op != "replace" {
+		return allErrs
+	}
+
+	if jsonPatch.Value == nil && jsonPatch.ValueFrom == nil {
+		allErrs = append(allErrs,
+			field.Invalid(
+				path,
+				prettyPrint(jsonPatch),
+				"jsonPatch must define one of value or valueFrom",
+			))
+	}
+
+	if jsonPatch.Value != nil && jsonPatch.ValueFrom != nil {
+		allErrs = append(allErrs,
+			field.Invalid(
+				path,
+				prettyPrint(jsonPatch),
+				"jsonPatch can not define both value and valueFrom",
+			))
+	}
+
+	// Attempt to marshal the JSON to discover if it  is valid. If jsonPatch.Value.Raw is set to nil skip this check
+	// and accept the nil value.
+	if jsonPatch.Value != nil && jsonPatch.Value.Raw != nil {
+		if err := json.Unmarshal(jsonPatch.Value.Raw, &struct{}{}); err != nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path,
+					string(jsonPatch.Value.Raw),
+					"jsonPatch Value is invalid JSON",
+				))
+		}
+	}
+	if jsonPatch.ValueFrom != nil && jsonPatch.ValueFrom.Template == nil && jsonPatch.ValueFrom.Variable == nil {
+		allErrs = append(allErrs,
+			field.Invalid(
+				path.Child("valueFrom"),
+				prettyPrint(jsonPatch.ValueFrom),
+				"valueFrom must set either template or variable",
+			))
+	}
+	if jsonPatch.ValueFrom != nil && jsonPatch.ValueFrom.Template != nil && jsonPatch.ValueFrom.Variable != nil {
+		allErrs = append(allErrs,
+			field.Invalid(
+				path.Child("valueFrom"),
+				prettyPrint(jsonPatch),
+				"valueFrom can not set both template and variable",
+			))
+	}
+
+	if jsonPatch.ValueFrom != nil && jsonPatch.ValueFrom.Template != nil {
+		// Error if template can not be parsed.
+		_, err := template.New("tpl").Parse(*jsonPatch.ValueFrom.Template)
+		if err != nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Child("valueFrom", "template"),
+					*jsonPatch.ValueFrom.Template,
+					fmt.Sprintf("template can not be parsed: %v", err),
+				))
+		}
+	}
+
+	// If set validate that the variable is valid.
+	if jsonPatch.ValueFrom != nil && jsonPatch.ValueFrom.Variable != nil {
+		// If the variable is one of the list of builtin variables it's valid.
+		if strings.HasPrefix(*jsonPatch.ValueFrom.Variable, "builtin") {
```

Comment:
```suggestion
		if strings.HasPrefix(*jsonPatch.ValueFrom.Variable, "builtin.") {
```
nit, variables like `builtinUserDefinedVariable` must be validated in the else